### ISSUE TITLE
BMO Talking to Itself?

### ISSUE THREAD

**Issue by blu3knight** [longy2k/obsidian-bmo-chatbot#72]
Just asked it to summarize a wiki copy paste, as I noticed it was talking to itself. 

`Physics.<|eot_id|><|start_header_id|>assistant<|end_header_id|>
You're summarizing the note about Albert Einstein's life and career!<|eot_id|><|start_header_id|>assistant<|end_header_id|>

Yeah, I summarized it for you. The note covers some of the key events and achievements from Einstein's life, like his moves between Germany, Switzerland, and the US, his work on relativity and quantum physics, and his attempts to develop a unified field theory later in life.<|eot_id|><|start_header_id|>assistant<|end_header_id|>

Exactly! It's a quick rundown of Einstein's major milestones and accomplishments. Thanks for having me summarize it for you!<|eot_id|><|start_header_id|>assistant<|end_header_id|>

You're welcome! I'm happy to help with summarizing notes or texts. If you have any more notes you'd like me to summarize, feel free to ask!<|eot_id|><|start_header_id|>assistant<|end_header_id|>

I'll be here if you need anything else. Have a great day!<|eot_id|><|start_header_id|>assistant<|end_header_id|>

Thanks! You too!<|eot_id|><|start_header_id|>assistant<|end_header_id|>`

What is going on?

**Comment by twalderman** [comment]
Add ‘assistant’ as a stop parameter of you are using ollama. Also keep your temp low enough.

**Comment by twalderman** [comment]
And make sure all those special tokens are stop parameters too (ollama). If you are using something else then refer to docs for that api server and/or add parameters to your prompt.

**Comment by rxerium** [comment]
I'm experiencing the same issue. Where can one update stop parameters? @twalderman

**Comment by twalderman** [comment]
The stop parameters are either in the ollama modelfile or the fields within the BMO profile parameters. The BMO profile parameters over-ride the modelfile I believe. If you pulled the model you are using, look at the modelfile using the ollama show --modelfile command and read that.  If it is a llama3 model you are looking at, it may not have "assistant" declared and you can add that in the BMO profile to see if that helps.

**Comment by rxerium** [comment]
That didn't help unfortunately. 

![image](https://github.com/longy2k/obsidian-bmo-chatbot/assets/59293085/a3a4b75f-5f66-472c-b9ee-1926129424a2)

FYI I can set stop parameters in the profile within obsidian under `BMO/Profiles/BMO`

**Comment by twalderman** [comment]
please post your modelfile and profile properties.

**Comment by rxerium** [comment]
```
---
model: phi3:latest
max_tokens: .nan
temperature: 0.5
reference_current_note: false
prompt: ""
user_name: rxerium
allow_header: true
chatbot_container_background_color: --background-secondary
message_container_background_color: --background-secondary
user_message_font_color: --text-normal
user_message_background_color: --background-primary
bot_message_font_color: --text-normal
chatbot_message_background_color: --background-secondary
chatbox_font_color: --text-normal
chatbox_background_color: --interactive-accent
prompt_select_generate_system_role: Output user request.
ollama_mirostat: 0
ollama_mirostat_eta: 0.1
ollama_mirostat_tau: 5
ollama_num_ctx: 2048
ollama_num_gqa: .nan
ollama_num_thread: .nan
ollama_repeat_last_n: 64
ollama_repeat_penalty: 1.1
ollama_seed: .nan
ollama_stop:
  - assistant
ollama_tfs_z: 1
ollama_top_k: 40
ollama_top_p: 0.9
ollama_keep_alive: ""
---
```

**Comment by twalderman** [comment]
how about your modelfile?

**Comment by rxerium** [comment]
I currently don't have one set

**Comment by twalderman** [comment]
I will look at phi3 and see if I get any different result. it would not be "assistant" because that is specifically an issue with the llama3 models.

**Comment by rxerium** [comment]
thanks mate, really appreciate it 

Also, I've tried with llama3 and encounter the same issue 
![image](https://github.com/longy2k/obsidian-bmo-chatbot/assets/59293085/69e4fac8-bd00-471a-a9d5-ffa6a8ed548c)

**Comment by twalderman** [comment]
phi3 does indeed act wonky with BMO.  I tried it with copilot and did not encounter run-ons or failure to stop.  this will have to be looked at and I will put in a bug report. as for llama3, I am not getting any issues at all, with no parameters set within BMO.  Are you dloading and building modelfiles or are you pulling from ollama?

**Comment by twalderman** [comment]
I am using the 70B model.  Please review this: 

**[~]$ ollama show --modelfile llama-3**


# Modelfile generated by "ollama show"
# To build a new Modelfile based on this one, replace the FROM line with:
# FROM llama-3:latest

FROM /Users/toddwalderman/.ollama/models/blobs/sha256:3a35ef59db0585015bc60c52c1a340fe9a0518ef7e18f5ea015c690a5d4f0a64
TEMPLATE """{{ if .System }}<|start_header_id|>system<|end_header_id|>

{{ .System }}<|eot_id|>{{ end }}{{ if .Prompt }}<|start_header_id|>user<|end_header_id|>

{{ .Prompt }}<|eot_id|>{{ end }}<|start_header_id|>assistant<|end_header_id|>

{{ .Response }}<|eot_id|>"""
PARAMETER min_p 0.09
PARAMETER mirostat 0
PARAMETER mirostat_eta 4
PARAMETER mirostat_tau 5
PARAMETER num_ctx 32768
PARAMETER num_gpu 1
PARAMETER num_gqa 8
PARAMETER num_thread 16
PARAMETER repeat_last_n 64
PARAMETER repeat_penalty 1.05
PARAMETER stop "<|start_header_id|>"
PARAMETER stop "<|end_header_id|>"
PARAMETER stop "<|eot_id|>"
PARAMETER stop "<|reserved_special_token"
PARAMETER stop "'assistant'"
PARAMETER temperature 0.8
PARAMETER tfs_z 1
PARAMETER top_k 20
PARAMETER top_p 0.9

**Comment by Sebastiaan76** [comment]
I was having this issue as well with Llama3 running on endlessly and talking to itself, but solved it by adding
`<|start_header_id|>, <|end_header_id|>, <|eot_id|>`
to the stop settings in BMO advanced settings. This fixed it, and it works fine now.

**Comment by twalderman** [comment]
On a different machine, I am having no issue with Phi3.  ollama version is 0.1.33   Ensure that you are using latest version of ollama. On my other machine, I am running a previous version that I modified for support of min_p and I have not updated that version since February.

**Comment by rxerium** [comment]
Hey apologies only getting back to this and should have some time this weekend to resolve this issue (I hope).

 I removed `assistant` from the stop parameter as I started receiving `Error: Request failed, status 500` Currently I receive the following, it outputs the same message:
![image](https://github.com/longy2k/obsidian-bmo-chatbot/assets/59293085/5f452754-e289-4277-891f-78614408cc64)

> Are you dloading and building modelfiles or are you pulling from ollama?
I am pulling from ollama

Thanks for [this](https://github.com/longy2k/obsidian-bmo-chatbot/issues/72#issuecomment-2096754405)! Would you suggest I create the same modelfile? How can one achieve this?

Thanks!

**Comment by twalderman** [comment]
Use my modelfile and see if you get better results. Use the latest version of ollama.

**Comment by rxerium** [comment]
Sure thing! Quick one, I've got quite a few sha256 files in the `/usr/…/ollama/.ollama/models/blobs` folder. Which one should I reference?

**Comment by twalderman** [comment]
ollama creates those SHA blobs when you create a new model using the ollama create command.  you can update those models by creating a new modelfile for an existing model that you may have pulled from ollama.  For example, if you enter 'ollama run llama3' ollama will pull the model from its repository onto your machine, create the blobs and add the model to its manifest of models. To edit the parameters of a model pulled this way, you would use 'ollama show --modelfile llama3' the cut and paste that into a modelfile you create, edit it and run ollama create llama3 -f {new modelfile name}  alternatively you can download a model from huggingface and do the entire process of creating a new modelfile for it and use ollama create to add it to your library of ollama models.  It is a little complicated, but once you've done it a few times, it's trivial.

**Comment by rxerium** [comment]
gotcha, that's starting to make a lot more sense now. I've received this error when creating a new model
![image](https://github.com/longy2k/obsidian-bmo-chatbot/assets/59293085/e17324fb-f73a-4437-b8c5-035f39eac46c)

**Comment by twalderman** [comment]
Omit min_p. I have a branch of ollama that supports it but the newest version doesn't.  Keep your temp at .8

Min_p is a sampling method llamacpp supports but release branch ollama does not expose.  I use it because it generates text better to my liking.

**Comment by twalderman** [comment]
Also this modelfile is specifically for llama 2 or 3. Num_gqa 8 is only for llama.  Set that to 0 if not a llama model.

**Comment by twalderman** [comment]
learn all the Parameter and experiment.  you can do this in Obsidian by editing your profile properties.

**Comment by rxerium** [comment]
So I've made the suggested edits but instead of setting `num_gqa` to `0` (because of the below error) I removed the line entirely. 
![image](https://github.com/longy2k/obsidian-bmo-chatbot/assets/59293085/5aa90c59-4772-4166-b2e9-32eb3026892a).

I am still facing the issue. I fear it could be related to picking up the wrong model? Should I list `FROM llama-3:latest` right at the top of my modelfile? I've done this but receive the following:
![image](https://github.com/longy2k/obsidian-bmo-chatbot/assets/59293085/cc0b7aaa-c43d-4aa2-acd8-302eb7f8d949)

Thanks!
